dual_net.py

```diff
@@ -0,0 +1,155 @@
+    - Convolutional layers:
+       - d['layer'] <- 'conv'
+       - d['num_outputs'], d['kernel_size'], d['stride'] should be ints
+       - d['activation_fn'] is a function
+    - Residual layers:
+       - d['layer'] <- 'residual'
+       - other keys same as convolutional
+    - Fully connected layers:
+       - d['layer'] <- 'fc'
+       - d['num_outputs'] is an int
+       - d['activation_fn'] is a function
+    policy_head and value_head have the same structure as above but represent
+    the layers for the policy head and value head, respectively.
+
+    returns the policy output and the value output in a tuple
+    """
+    out = board_placeholder
+    for specs in shared_layers:
+        if specs['layer'] == 'conv':
+            out = conv_block(out, specs)
+        elif specs['layer'] == 'residual':
+            out = residual_block(out, specs)
+        elif specs['layer'] == 'fc':
+            out = layers.flatten(out)
+            out = layers.fully_connected(out,
+                                         num_outputs=specs['num_outputs'],
+                                         activation_fn=specs['activation_fn'])
+    # Policy head
+    policy_out = out
+    for specs in policy_head:
+        if specs['layer'] == 'conv':
+            policy_out = conv_block(policy_out, specs)
+        elif specs['layer'] == 'fc':
+            policy_out = layers.flatten(policy_out)
+            policy_out = layers.fully_connected(policy_out,
+                                                num_outputs=specs['num_outputs'],
+                                                activation_fn=specs['activation_fn'])
+    policy_out = tf.nn.log_softmax(policy_out)
+
+    # Value head
+    value_out = out
+    for specs in value_head:
+        if specs['layer'] == 'conv':
+            value_out = conv_block(value_out, specs)
+        elif specs['layer'] == 'fc':
+            value_out = layers.flatten(value_out)
+            value_out = layers.fully_connected(value_out,
+                                               num_outputs=specs['num_outputs'],
+                                               activation_fn=specs['activation_fn'])
+    return policy_out, value_out
+
+
+class DualNet(object):
+
+    def __init__(self, sess, action_rep='pos', learning_rate=0.01, regularization_mult=0.01):
+        """
+        sess: tensorflow session
+        action_rep: 'pos' for (pos1, pos2) action representation, 'piece' for (piece, pos) rep
+        learning_rate: learning rate for gradient descent
+        regularization_mult: multiplier for weight regularization loss
+        """
+        self.board_placeholder = tf.placeholder(tf.float32, [None, 8, 8, 13])
+        shared_layers = [{'layer': 'conv', 'num_outputs': 256, 'stride': 3, 'kernel_size': 1, 'activation_fn': tf.nn.relu},
+                         {'layer': 'residual', 'num_outputs': 256, 'stride': 1, 'kernel_size': 3, 'activation_fn': tf.nn.relu},
```

Comment:
To clarify, they have 20 or 40 residual layers right? Are we just starting with two for now?

Reply:
Yeah Josh recommended starting with a smaller net to see how well that performs.

Reply:
Nice yeah I agree! I say we add that bad boy in as a parameter - I'll push something with it